pkg/ds/queue/queue.go

```diff
@@ -0,0 +1,155 @@
+package queue
+
+type Node[T any] struct {
+	value T
+	next  *Node[T]
+	prev  *Node[T]
+}
+
+type Queue[T any] struct {
+	head *Node[T]
+	tail *Node[T]
+	size uint
+}
+
+func New[T any]() *Queue[T] {
+	return &Queue[T]{
+		head: nil,
+		tail: nil,
+		size: 0,
+	}
+}
+
+func (q *Queue[T]) PushBack(val T) {
+	q.size++
+	node := &Node[T]{
+		value: val,
+		next:  nil,
+		prev:  nil,
+	}
+
+	if q.head == nil {
+		q.head = node
+		q.tail = node
+		return
+	}
+
+	node.prev = q.tail
+	q.tail.next = node
+	q.tail = node
+}
+
+func (q *Queue[T]) PushFront(val T) {
+	q.size++
+	node := &Node[T]{
+		value: val,
+		next:  nil,
+		prev:  nil,
+	}
+
+	if q.head == nil {
+		q.head = node
+		q.tail = node
+		return
+	}
+
+	node.next = q.head
+	q.head.prev = node
+	q.head = node
+}
+
+func (q *Queue[T]) PopBack() (T, bool) {
+	var t T
+
+	// 0 node
+	if q.tail == nil {
+		return t, false
+	}
+
+	q.size--
+	t = q.tail.value
+	// only 1 node
+	if q.tail == q.head {
+		q.tail = nil
+		q.head = nil
+		return t, true
+	}
+
+	if q.tail.prev != nil {
+		q.tail.prev.next = nil
+	}
+	q.tail = q.tail.prev
+
+	return t, true
+}
+
+func (q *Queue[T]) PopFront() (T, bool) {
+	var t T
+
+	// 0 node
+	if q.head == nil {
```

Comment:
```suggestion
	if q.IsEmpty() {
```

Reply:
updated in https://github.com/KyberNetwork/tradinglib/pull/13/commits/b77b7dc94ee7c6d4ae2eeaf97ac3686310d30d5e